indico/modules/rb_new/client/js/components/TimelineItem.jsx

```diff
@@ -0,0 +1,124 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Indico; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+import PropTypes from 'prop-types';
+import React from 'react';
+import {Popup, Segment} from 'semantic-ui-react';
+import {Translate} from 'indico/react/i18n';
+
+import './TimelineItem.module.scss';
+
+
+const colorMapping = {
+    candidates: 'green',
+    bookings: 'orange',
+    conflicts: 'red',
+};
+
+const classes = {
+    candidates: 'candidate',
+    preBookings: 'pre-booking',
+    preConflicts: 'pre-booking-conflict'
+};
+
+
+export default class TimelineItem extends React.Component {
+    static propTypes = {
+        startHour: PropTypes.number.isRequired,
+        endHour: PropTypes.number.isRequired,
+        step: PropTypes.number.isRequired,
+        data: PropTypes.object.isRequired
+    };
+
+    calculateWidth = (startDt, endDt) => {
+        const {startHour, endHour, step} = this.props;
+        let segmentStartDuration = (startDt.hours() * 60) + startDt.minutes();
+        let segmentEndDuration = (endDt.hours() * 60) + endDt.minutes();
+
+        if (segmentStartDuration < startHour * 60) {
+            segmentStartDuration = startHour * 60;
+        }
+        if (segmentEndDuration > endHour * 60) {
+            segmentEndDuration = endHour * 60;
+        }
+
+        return (segmentEndDuration - segmentStartDuration) / (step * 60);
+    };
+
+    calculatePosition = (startDt) => {
+        const {startHour, step} = this.props;
+        let segmentStartDuration = (startDt.hours() * 60) + startDt.minutes();
+        if (segmentStartDuration < startHour * 60) {
+            segmentStartDuration = startHour * 60;
+        }
+        return (segmentStartDuration - (startHour * 60)) / (step * 60);
+    };
+
+    renderOccurrence = (occurrence, color = null, additionalClasses = '') => {
+        const {start_dt: startDt, end_dt: endDt, reservation} = occurrence;
+        const {startHour, endHour, step} = this.props;
+        const segmentStartDt = moment(startDt, 'YYYY-MM-DD HH:mm');
+        const segmentEndDt = moment(endDt, 'YYYY-MM-DD HH:mm');
+        const blockWidth = 100 / (((endHour - startHour) / step) + 1);
+        const segmentWidth = this.calculateWidth(segmentStartDt, segmentEndDt) * blockWidth;
+        const segmentPosition = this.calculatePosition(segmentStartDt) * blockWidth;
+
+        const popupContent = (
+            <>
+                <div>
+                    {segmentStartDt.format('HH:mm')} - {segmentEndDt.format('HH:mm')}
+                </div>
+                <div>
+                    {reservation ? reservation.booking_reason : Translate.string('Click to book it')}
+                </div>
+            </>
+        );
+        const segment = (
+            <Segment color={color ? color : null} className={additionalClasses}
```

Comment:
Is `<Segment />` really the best option? The component doesn't seem to do what we intend here...

Reply:
Yeah, it might be a little bit misused (but did the job at the beginning, I am gonna change it anyway)